### PR TITLE
Explicit broadcast and testing

### PR DESCRIPTION
**Pull request by OwenArnold**
This is not to be considered until AFTER the demo sessions.

The changes here are essentially cleanup.

### CODE REVIEW

File: test/dataset_test.cpp [hunk 1]
```diff
@@ -574,14 +574,20 @@ TEST(Dataset, operator_multiplication_broadcast) {
...
   a.insert(Data::Variance, Dimensions({{Dim::X, 2}, {Dim::Y, 2}}),
            {1.0, 1.0, 1.0, 1.0});
 
   Dataset b;
   b.insert(Data::Value, {Dim::Y, 2}, {2.0, 3.0});
+  b(Data::Value).setUnit(units::m);
   b.insert(Data::Variance, {Dim::Y, 2}, {1.0, 1.0});
 
   auto c = a * b;
+  // Check units
```
Comment (nvaytet) [inline test/dataset_test.cpp]: Does it work with `a / a` ? We were having problems with dimensionless not being the same as `si::dimensionless`

Comment (OwenArnold) [reply test/dataset_test.cpp]: @nvaytet I've just added a new test. S

Comment (OwenArnold) [reply test/dataset_test.cpp]: Works for `c = a / b` too.

Comment (nvaytet) [reply test/dataset_test.cpp]: Canout you also try with `units.counts`? Because we were having problems with `counts^2 * counts^2`

Comment (OwenArnold) [reply test/dataset_test.cpp]: Was your problem specifically around examples where there is dimension mismatch between rhs and lhs arguments? The changes here relate only to broadcast - I'm happy to check though.

Comment (OwenArnold) [reply test/dataset_test.cpp]: So extending the unit definitions to include `counts ^ 4` and switching units to `counts` ensures that both these new test methods pass. However, note this is specifically tested against for failure [here](https://github.com/mantidproject/dataset/blob/master/test/dataset_test.cpp#L836), which makes sense

**Pull request merged**